### PR TITLE
feat: `CountingSort` implementation

### PR DESCRIPTION
**Pull request by alxkm**
CountingSort implementation.

Previous discussion in this PR
https://github.com/TheAlgorithms/Java/pull/5275

- [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Java/blob/master/CONTRIBUTING.md).
- [x] This pull request is all my own work -- I have not plagiarized it.
- [x] All filenames are in PascalCase.
- [x] All functions and variable names follow Java naming conventions.
- [x] All new algorithms have a URL in their comments that points to Wikipedia or other similar explanations.
- [x] All new code is formatted with `clang-format -i --style=file path/to/your/file.java`

### CODE REVIEW

File: src/main/java/com/thealgorithms/sorts/CountingSort.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+/**
+ * A standard implementation of the Counting Sort algorithm for integer arrays.
+ * This implementation has a time complexity of O(n + k), where n is the number
+ * of elements in the input array and k is the range of the input.
+ * It works only with integer arrays.
+ *
+ * The space complexity is O(k), where k is the range of the input integers.
+ *
+ * Note: This implementation does not handle negative integers as it
+ * calculates the range based on the minimum and maximum values of the array.
```
Comment (vil02) [inline src/main/java/com/thealgorithms/sorts/CountingSort.java:14]: I think your implementation indeed works for negative numbers.

By the way: to inform the user by throwing an exception for example.

File: src/main/java/com/thealgorithms/sorts/CountingSort.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+     * @param array the array to be sorted
+     * @return the sorted array
+     */
+    public static int[] sort(int[] array) {
+        if (array.length == 0) {
+            return array;
+        }
+
+        final int max = Arrays.stream(array).max().orElse(Integer.MIN_VALUE);
+        final int min = Arrays.stream(array).min().orElse(Integer.MAX_VALUE);
```
Comment (vil02) [inline src/main/java/com/thealgorithms/sorts/CountingSort.java:33]: At this point `array` cannot be empty, therefore, I would suggest to use `getAsInt()`.

File: src/main/java/com/thealgorithms/sorts/CountingSort.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+            count[i] += count[i - 1];
+        }
+
+        for (int i = array.length - 1; i >= 0; i--) {
+            output[count[array[i] - min] - 1] = array[i];
+            count[array[i] - min]--;
+        }
+
+        return output;
+    }
```
Comment (vil02) [inline src/main/java/com/thealgorithms/sorts/CountingSort.java:53]:
```suggestion
    public static int[] sort(int[] array) {
        if (array.length == 0) {
            return array;
        }

        final var stats = Arrays.stream(array).summaryStatistics();
        final int min = stats.getMin();

        int[] count = computeHistogram(array, min, stats.getMax() - min + 1);
        toCumulative(count);
        return reconstructSorted(count, min, array);
    }

    private static int[] computeHistogram(final int[] array, final int shift, final int spread) {
        int[] res = new int[spread];
        for (final var value : array) {
            res[value - shift]++;
        }
        return res;
    }

    private static void toCumulative(int[] count) {
        for (int i = 1; i < count.length; i++) {
            count[i] += count[i - 1];
        }
    }

    private static int[] reconstructSorted(final int[] cumulativeCount, final int shift, final int[] array) {
        int[] res = new int[array.length];
        for (int i = array.length - 1; i >= 0; i--) {
            res[cumulativeCount[array[i] - shift] - 1] = array[i];
            cumulativeCount[array[i] - shift]--;
        }

        return res;
    }
```

File: src/test/java/com/thealgorithms/sorts/CountingSortTest.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+public class CountingSortTest {
+
+    record TestCase(int[] inputArray, int[] expectedArray) {
+    }
+
+    static Stream<TestCase> provideTestCases() {
+        return Stream.of(new TestCase(new int[] {}, new int[] {}), new TestCase(new int[] {4}, new int[] {4}), new TestCase(new int[] {6, 1, 99, 27, 15, 23, 36}, new int[] {1, 6, 15, 23, 27, 36, 99}), new TestCase(new int[] {6, 1, 27, 15, 23, 27, 36, 23}, new int[] {1, 6, 15, 23, 23, 27, 27, 36}),
+            new TestCase(new int[] {5, 5, 5, 5, 5}, new int[] {5, 5, 5, 5, 5}), new TestCase(new int[] {1, 2, 3, 4, 5}, new int[] {1, 2, 3, 4, 5}), new TestCase(new int[] {5, 4, 3, 2, 1}, new int[] {1, 2, 3, 4, 5}), new TestCase(new int[] {3, -1, 4, 1, 5, -9}, new int[] {-9, -1, 1, 3, 4, 5}),
+            new TestCase(new int[] {0, 0, 0, 0}, new int[] {0, 0, 0, 0}), new TestCase(new int[] {3, 3, -1, -1, 2, 2, 0, 0}, new int[] {-1, -1, 0, 0, 2, 2, 3, 3}), new TestCase(new int[] {-3, -2, -1, -5, -4}, new int[] {-5, -4, -3, -2, -1}),
+            new TestCase(new int[] {1000, 500, 100, 50, 10, 5, 1}, new int[] {1, 5, 10, 50, 100, 500, 1000}));
```
Comment (vil02) [inline src/test/java/com/thealgorithms/sorts/CountingSortTest.java:18]:
```suggestion
            new TestCase(new int[] {1000, 500, 100, 50, 10, 5, 1}, new int[] {1, 5, 10, 50, 100, 500, 1000}), new TestCase(new int[] {4, -5, 10, 0}, new int[] {-5, 0, 4, 10}));
```

**Review by vil02 — CHANGES_REQUESTED** [review]

File: src/main/java/com/thealgorithms/sorts/CountingSort.java (hunk 1 above)
Comment (alxkm) [reply src/main/java/com/thealgorithms/sorts/CountingSort.java:14]: Yes, fixed this. Thanks.

File: src/main/java/com/thealgorithms/sorts/CountingSort.java (hunk 3 above)
Comment (alxkm) [reply src/main/java/com/thealgorithms/sorts/CountingSort.java:53]: Thanks. Such structure easier to understand

File: src/main/java/com/thealgorithms/sorts/CountingSort.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+/**
+ * A standard implementation of the Counting Sort algorithm for integer arrays.
+ * This implementation has a time complexity of O(n + k), where n is the number
+ * of elements in the input array and k is the range of the input.
+ * It works only with integer arrays.
+ *
+ * The space complexity is O(k), where k is the range of the input integers.
+ *
+ * Note: This implementation handle negative integers as it
```
Comment (vil02) [inline src/main/java/com/thealgorithms/sorts/CountingSort.java:13]:
```suggestion
 * Note: This implementation handles negative integers as it
```

**Review by vil02 — APPROVED** [review]

**Pull request merged**